src/server/scripts/Spells/spell_priest.cpp

```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
         }
 };
 
+//17 - Power Word : Shield
+class spell_pri_power_word_shield : public SpellScriptLoader
+{
+public:
+    spell_pri_power_word_shield() : SpellScriptLoader("spell_pri_power_word_shield") { }
+
+    enum spell_enum
+    {
+        SPELL_PRIEST_BODY_AND_SOUL              = 64129,
+        SPELL_PRIEST_BODY_AND_SOUL_ACTIVE       = 65081,
+        SPELL_PRIEST_SHIELD_DISCIPLINE_PASSIVE  = 197045,
+        SPELL_PRIEST_SHIELD_DISCIPLINE_ENERGIZE = 47755,
+        SPELL_PRIEST_STRENGHT_OF_SOUL           = 197535,
+        SPELL_PRIEST_STRENGHT_OF_SOUL_EFFECT    = 197548,
+        SPELL_PRIEST_RENEWED_HOPE               = 197469,
+        SPELL_PRIEST_RENEWED_HOPE_EFFECT        = 197470,
+        SPELL_PRIEST_VOID_SHIELD                = 199144,
+        SPELL_PRIEST_VOID_SHIELD_EFFECT         = 199145
+    };
+    class spell_pri_power_word_shield_SpellScript : public SpellScript
+    {
+
+        PrepareSpellScript(spell_pri_power_word_shield_SpellScript);
+
+        void HandleAfterCast()
+        {
+            Player* caster = GetCaster()->ToPlayer();
+
+            if (caster->HasAura(SPELL_PRIEST_BODY_AND_SOUL))
+                caster->CastSpell(caster, SPELL_PRIEST_BODY_AND_SOUL_ACTIVE,true);
+            if (caster->HasAura(SPELL_PRIEST_VOID_SHIELD) && GetHitUnit()->GetGUID() == caster->GetGUID())
+                caster->CastSpell(caster, SPELL_PRIEST_VOID_SHIELD_EFFECT, true);
+            if (caster->HasAura(SPELL_PRIEST_STRENGHT_OF_SOUL))
+                caster->CastSpell(caster, SPELL_PRIEST_STRENGHT_OF_SOUL_EFFECT,true);
+            if (caster->HasAura(SPELL_PRIEST_RENEWED_HOPE))
+                caster->CastSpell(caster, SPELL_PRIEST_RENEWED_HOPE_EFFECT, true);
+        }
+
+        void Register() override
+        {
+            AfterCast += SpellCastFn(spell_pri_power_word_shield_SpellScript::HandleAfterCast);
+        }
+    };
+
+    class spell_pri_power_word_shield_AuraScript :public AuraScript
+    {
+        PrepareAuraScript(spell_pri_power_word_shield_AuraScript);
+
+        void CalculateAmount(const AuraEffect* /*auraEffect*/, int32& amount, bool& /*canBeRecalculated*/)
+        {
+            Player* caster = GetCaster()->ToPlayer();
+            int32 sp = caster->SpellBaseDamageBonusDone(SPELL_SCHOOL_MASK_HOLY);
+            int32 versadmg = caster->GetRatingBonusValue(CombatRating::CR_VERSATILITY_DAMAGE_DONE);
+            int32 mastery = caster->GetRatingBonusValue(CombatRating::CR_MASTERY) + 1;
+            amount = (int32)((sp * 5.5f) + mastery) * (1 + versadmg);
+        }
+        
+        void Absorb(AuraEffect* aurEff, DamageInfo & dmgInfo, uint32 & absorbAmount)
```

Comment:
this function is not crash safe, make sure GetEffect is checked and ToPlayer is checked, also rename "n" and "m" to something easier to understand. 
Another suggestion:

```c++
if (n <= m)
    absorbAmount += (int32)dmgInfo.GetDamage();
```